### PR TITLE
More exceptions

### PR DESCRIPTION
**Pull request by mirceaulinic**
Define new exception `ConnectionClosedException` as per https://github.com/napalm-automation/napalm-base/pull/247

Add the exceptions proposed in https://github.com/napalm-automation/napalm-base/issues/218

### CODE REVIEW

File: napalm_base/exceptions.py [hunk 1]
```diff
@@ -22,6 +22,35 @@ class ModuleImportError(Exception):
 
 
 class ConnectionException(Exception):
+    '''
+    Unable to connect to the network device.
+    Raised by the ``open`` method.
```
Comment (dbarrosop) [inline napalm_base/exceptions.py]: This last sentence is not accurate. For two reasons;

1. childs of this class will be thrown on other places, which means you can actually catch this exception in other methods.
2. It kind of implies this is the only exception that is going to be thrown by the open method while the truth is that it would be nice if its childs were raised instead.

File: napalm_base/exceptions.py [hunk 2]
```diff
@@ -22,6 +22,35 @@ class ModuleImportError(Exception):
...
+    '''
+    Unable to connect to the network device
+    due to invalid credentials.
+    '''
+    pass
+
+
+class ConnectTimeoutError(ConnectionException):
+    '''
+    Connection to the network device takes too long.
```
Comment (dbarrosop) [inline napalm_base/exceptions.py]: Maybe mention timeout value can be tweaked with the timeout argument. Mostly to indicate there is a correlation.

File: napalm_base/exceptions.py [hunk 3]
```diff
@@ -22,6 +22,35 @@ class ModuleImportError(Exception):
...
+    '''
+    pass
+
+
+class ConnectionClosedException(ConnectionException):
+    '''
+    The network device closed the connection.
```
Comment (dbarrosop) [inline napalm_base/exceptions.py]: I would reword this and use "dropped' instead unless we are expecting to raise this exception if the server ends the TCP socket. My guess is we will be raising this if the connection is silently dropped by a firewall or the server.

**Comment by mirceaulinic** [comment]
@dbarrosop I've pushed some changes: https://github.com/napalm-automation/napalm-base/pull/248/commits/b3b5c9be7982169f43968fa5af9291e8af1c2807, feel free to suggest more improvements if anything still doesn't sound good.

**Comment by ktbyers** [comment]
In SSH you can also have the SSH channel closed by remote end (currently raises an EOFError in Netmiko). In this case the TCP connection is still open, but you are unable to communicate.

I think this should probably fall into this `ConnectionClosedException` as you would need to treat it the same way in practice (i.e. re-open a new connection).

**Comment by mirceaulinic** [comment]
Yes, that was my understanding too @ktbyers. Would you suggest including this explanation in the exception docsting?

**Comment by ktbyers** [comment]
I think we might want to include that in the docstring (i.e. 'or SSH channel closing').

**Comment by dbarrosop** [comment]
We don't really need super explicit. Most of the time people will just want to know the connection was dropped due to inactivity (regardless of who or how) and just try to reconnect. My point is that a single ConnectionDroppedException and then different messages for logging purposes should probably be enough. I can't imagine someone treating a timeout because a firewall dropped a flow being treated different than ssh closing the channel.
